src/msg/async/AsyncConnection.cc

```diff
@@ -967,17 +967,31 @@ int AsyncConnection::_process_connection()
           ::close(sd);
         }
 
-        sd = net.connect(get_peer_addr());
+        sd = net.nonblock_connect(get_peer_addr());
         if (sd < 0) {
           goto fail;
         }
-        r = net.set_nonblock(sd);
+        center->create_file_event(sd, EVENT_READABLE, read_handler);
+        first_connect = true;
+        state = STATE_CONNECTING_RE;
+        break;
+      }
+
+    case STATE_CONNECTING_RE:
+      {
+        r = net.reconnect(get_peer_addr(), sd);
         if (r < 0) {
+          if (first_connect == true) {
```

Comment:
When first call nonblock_connect, if we call reconnect fail. We don't need to go to fail and retry, we just break. In other cases, we must retry.
When you have many breakdown peer, retrying may make msg thread busy, and also mark_down other peer wrongly.